base/associative.jl

```diff
@@ -136,8 +136,42 @@ This includes arrays, where the keys are the array indices.
 """
 pairs(collection) = Generator(=>, keys(collection), values(collection))
 
+"""
+    similar(a::Associative, [value_type=valtype(a)], [inds=keys(a)])
+
+Create an `Associative` container with unitialized values of `value_type` and indices
+`inds`. The second and third arguments are optional and default to the input's `valtype` and
+`keys`, respectively.
+
+Custom `Associative` subtypes may choose which specific associative type is best suited to
+return for the given value type and indices, by specializing on the three-argument
+signature. The default is to return a `Dict`.
+"""
+# v1.0: similar(a::Associative) = similar(a, valtype(a), keys(a))
+similar(a::Associative, ::Type{T}) where {T} = similar(a, T, keys(a))
+similar(a::Associative, inds) = similar(a, valtype(a), inds)
+
+# disambiguation
+similar(a::Associative, t::Tuple) = similar(a, eltype(a), t)
```

Comment:
Did you mean `valtype(a)` instead of `eltype(a)`?

Also, doesn't this mean that `similar(::Array, (2,3))` returns a 2x3 array, but `similar(::Dict, (2,3))` returns a dict with keys 2 and 3?

Reply:
Ah, yes, thanks :)

> Also, doesn't this mean that `similar(::Array, (2,3))` returns a 2x3 array, but `similar(::Dict, (2,3))` returns a dict with keys 2 and 3?

That's right, and based on thinking about how to make arrays and dicts play together nicely that's something I'd prefer to see changed. :) I see the varargs versions of `similar` making more sense here, so I would propose that

```
similar(array, CartesianRange(2,3))
similar(array, 2, 3)
similar(array, OneTo(2), OneTo(3))
```

are all the same and that `similar(array, (2,3))` be deprecated and maybe in the future create something with `keys` of `2` and `3` (well that was what I did in [AssociativeArray](https://github.com/andyferris/AssociativeArray.jl) - useful where you might want to have an associative "similar" to an `DArray` or an `SArray`, or vice-versa).

From here I'd define something like `similar(array, dims_or_inds...) = similar(array, cartesian(dims_or_inds...))` and `cartesian(dims::Integer...) = CartesianRange(dims)` and similarly for `cartesian(inds::OneTo...)` and users can overload this for offset arrays and their own conversions between their custom unit range / Cartesian range types. One might consider the same trick for transforming varargs `getindex(array, x, y)`, in order to transform to `getindex(array, key)` with `key in keys(array)` (for example `key = CartesianIndex(x, y)` if `x` and `y` are `Integer`, or else a `Slice` for non-scalar multidimensional indexing).

Then, a user could implement varargs versions of `getindex` and `similar` for `Associative` if they wanted to enable Cartesian indexing of an `Associative` (is this maybe like an AxisArray?). To make that happen we'd also need to get rid of the tuple stuff from `getindex(associative, inds...)` and leave it to packages to define what that means. (User's may still of course use a tuple as keys of a `Dict` directly)

Reply:
(fixed and relevant tests added)